Input fixes/improvements

RobertTheSable opened the pull request:
- Add L3/R3 as aliases for soft shoulder presses for GC Controllers.
- Allow more range when analog triggers are used for GC Controllers.
- Enable GC Controller input in ports 5-8 in Wii mode.
- Fix sideways Wiimote input expressions so the input descriptors are correct.

Commit:
Add L/R Analog Input options for GCPads.

Commit:
Various input fixes.
- Add L3/R3 as alias for soft shoulder presses.
- Allow more range when analog triggers are used for GC Controllers.
- Enable GC Controller input in ports 5-8 in Wii mode.
- Fix sideways Wiimote input expressions.
- Fix tilt orientation in sideways wiimotes.
- Add tilt controls to Wiimote + Nunchuck.
- Alias ports 5-8 and 1-4 in Wii mode.

i30817 commented:
Hey is it my imagination or does this core have a problem with the 'start' key being switched with the 'select' key?

I started the game 'Eternal Darkness' with a retropad mapped to the keyboard and noticed that the menu couldn't be entered (well sort of, Y can enter it for enchanting only). 

The menu in this game is supposed to be 'start', and 'start' was enter and 'select' was 'right shift'. This shows in the game/core inputs that 'select' is unmapped (because the gamecube controller doesn't have it) and Start is Enter.

But the game menu will only be available if 'Select button' (that is right shift in the keyboard) is mapped to 'Start'. If i use 'Start button' (mapped to enter in the keyboard) nothing happens even if it's mapped to 'Start' too.

This is just from core buildbot origin, i don't have the PR built.

bslenul commented:
I messed a lot with f4f4947 (latest on master), 1fbd599 (latest available on Windows core updater, before the RA hack), 2dd20d2 (latest available on Linux core updater) and this PR here, never encountered that during my testings, weird... 🤔

i30817 commented:
the name that appears when you load the core is  1.9.0 - dolphin-emu (2dd20d2), so it's either a very old bug or RA itself screwed up? I downloaded the core from the updater. Should I have been building it?


I'm specifically using a keyboard as a 'retropad' and mapped Enter to start and Select to right shift. Then when i load that dolphin and assign the core input to 'gc controller' see that select is empty (because there is no select in that controller) and start is enter, then in game i can only open the menu if i map select to start too, because the enter key does nothing.

i30817 commented:
The only thing that is likely to be different is that i'm using a master build RA (with wayland support instead of X11). I tried to reset all settings by deleting the settings/dolphin-emu dir and the retroarch.cfg and the retroarch-core-settings.cfg and it still had the same problems.

I'm going to try to update/build the core before going to report problems in RA.

inactive123 commented:
@aliaspider You'd like to review this?

i30817 commented:
> I messed a lot with [f4f4947](https://github.com/libretro/dolphin/commit/f4f49478b602ae96bab7389788eaa43a2ab00fbd) (latest on master), [1fbd599](https://github.com/libretro/dolphin/commit/1fbd59911d1b718c142d6448dee3ede98152e395) (latest available on Windows core updater, before the RA hack), [2dd20d2](https://github.com/libretro/dolphin/commit/2dd20d2c016517646010c28bba2a0814a55461ef) (latest available on Linux core updater) and this PR here, never encountered that during my testings, weird... thinking

I think it's a retroarch problem, i've found that mapping enter to start doesn't work in *any* core, and opened a issue.

bslenul commented:
Hey!

I never really tried Wii games with the core until today, so I just noticed that by default it considers that NO controller is plugged at all when launching a Wii title.

So I was wondering, is there a way to detect when it's a Wii game and replace the default (empty) "RetroPad" with something else (Wiimote + Nunchuk maybe? Or just Wiimote? I honestly don't know which one would make the most sense).

Or maybe just restore "GameCube Controller" by default, even if it's a Wii game, so it's less confusing for the user (+ GC controller is compatible with some Wii titles like Mario Kart!). I mean if the user sees that it's not working and sees "GameCube Controller" in Controls menu it will make sense for him to why it's not working.

ds22x commented:
I second bslenul's suggestion to restore the GameCube controller instead of the retropad as default for Wii games.

I would also like to ask if it is possible to add to more entries for the GameCube controllers L and R analog, except without the full press when fully pressing the button.
That would allow for D-Input controllers to access the analog and (already available) full press by setting them to separate buttons (something that's already possible with standalone Dolphin).
I would also like to ask if it is possible to add Wiimote tilt to the Wiimote and Nunchuk profile.

Besides that, I've been testing this PR for a week now, and it seems to work just fine, except for one issue regarding the sideways Wiimote profile.
It seems the tilt values are still configured as if the Wiimote is held upright, as whenever I get to a point in New Super Mario Bros. Wii where tilting is required, it immediately tilts to the right without any input required.
Using the upright Wiimote profile, tilting works just fine.

bslenul commented:
> I would also like to ask if it is possible to add to more entries for the GameCube controllers L and R analog, except without the full press when fully pressing the button.

With this PR it's already possible, it's defaulted to L3 and R3:

![image](https://user-images.githubusercontent.com/33353403/95462509-e1dc4700-0977-11eb-9845-8743cf8a0cca.png)

So basically on keyboard or digital controller, pressing R2 will do a full press, and R3 a half press.

ds22x commented:
Well that's weird.
When I use it, L analog and R analog work exactly as I previously described it, namely as analog with at the end of the spring a full press.
Or is it that it only separates the full presses whenever you set them to another input?

RobertTheSable commented:
> Well that's weird.
> When I use it, L analog and R analog work exactly as I previously described it, namely as analog with at the end of the spring a full press.
> Or is it that it only separates the full presses whenever you set them to another input?

If you map l/r-analog to a separate digital button, pressing that should register only as a soft press. 

FYI: There appears to be a bug currently in Retroarch where analog buttons on the controller can't be remapped properly. I.E. if you remap L2 from L->something else, it still registers as L in the core. It sounds like this might be what's affecting you. AFAIK there's nothing I can do about it in the scope of this PR while still supporting analog input. A workaround I've found is to remap your buttons in the global controls setting rather than in the core settings.

ds22x commented:
I see.
Well that's unfortunate, but at least I got that cleared up for myself.
I'll just have to readjust my controls accordingly.

RobertTheSable commented:
> Hey!
> 
> I never really tried Wii games with the core until today, so I just noticed that by default it considers that NO controller is plugged at all when launching a Wii title.
> 
> So I was wondering, is there a way to detect when it's a Wii game and replace the default (empty) "RetroPad" with something else (Wiimote + Nunchuk maybe? Or just Wiimote? I honestly don't know which one would make the most sense).
> 
> Or maybe just restore "GameCube Controller" by default, even if it's a Wii game, so it's less confusing for the user (+ GC controller is compatible with some Wii titles like Mario Kart!). I mean if the user sees that it's not working and sees "GameCube Controller" in Controls menu it will make sense for him to why it's not working.

Wii mode having nothing plugged in by default is because the default controller is the retropad, and retropad isn't mapped to anything in Wii mode(except on ports 5-8 in my PR)

While I was working on the PR I was able to allow GC controllers in all 8 ports, with port 1 and 5, 2 and 6, and so on mapping to the same controller in the core. I wasn't sure if this was the desired behavior, so I didn't go through with it, but I could reimplement that.

Obviously the ideal would be to have up to 4 wii and 4 gc controllers in any port, but I'm not sure if there's an ideal way to represent that in RA.

RobertTheSable commented:
I think that last commit should resolve the issues with sideways and nunchuck wiimote tilt.

For the default device issue - I have a solution now which sets Wiimote as the default device. However based on discussion here, I think mirroring the GC controllers was the preferred solution? If that's the case I'll need some time to figure out a solution I'm happy with. I know I said I did it that way before, but I forget how I implemented it - I think I used control expressions, but I find that solution kind of messy, so I might try to see if I can find a better one.

RobertTheSable commented:
I'm unsure if the rumble code I wrote works because it turns out that I don't own any controllers with rumble(or at least that support rumble when used in Linux), so if someone who does wants to try it, that would be appreciated.

aliaspider commented:
making a PR with a cherry-picked commit, that was deliberately left out of the rebase, doesn't really help anyone, because even if I were to review it, you wouldn't know how to answer or make the necessary changes.

the commit in question had several issues, mostly unnecessary refactors and changes to core code that break the standalone build.

bslenul commented:
> I'm unsure if the rumble code I wrote works because it turns out that I don't own any controllers with rumble(or at least that support rumble when used in Linux), so if someone who does wants to try it, that would be appreciated.

Rumble works for me, with or without your PR.

For default controller, IMO Wiimote or Wiimote + Nunchuk would make more sense when booting Wii games, I mean on the real console I doubt many people used anything else on boot :p But I'm not a dev so it's not my call of course.

As of right now with your latest changes, it defaults to GC controller for Wii games in the controls menu, but it still acts like no controller is plugged at all. For example a GC controller should work with Mario Kart Wii but it doesn't.

RobertTheSable commented:
> > I'm unsure if the rumble code I wrote works because it turns out that I don't own any controllers with rumble(or at least that support rumble when used in Linux), so if someone who does wants to try it, that would be appreciated.
> 
> Rumble works for me, with or without your PR.
> 
> For default controller, IMO Wiimote or Wiimote + Nunchuk would make more sense when booting Wii games, I mean on the real console I doubt many people used anything else on boot :p But I'm not a dev so it's not my call of course.
> 
> As of right now with your latest changes, it defaults to GC controller for Wii games in the controls menu, but it still acts like no controller is plugged at all. For example a GC controller should work with Mario Kart Wii but it doesn't.

Does it behave different in Port 1 vs. Port 5?

For reference, I've mostly been using this for controller testing in Wii mode: https://wiibrew.org/wiki/Visual_Controller_Test

Well, also Bomberman Blast, since AFAIK that's the only game that uses the Wii and Gamecube controllers simultaneously.

bslenul commented:
> Does it behave different in Port 1 vs. Port 5?

Ah yep, it works if I set my controller to port 5. Is that on purpose then? Seems a bit inconvenient if you have "Max Users" set to "4" or less in RetroArch :p

aliaspider commented:
it does look a lot better now :)

I think you need a new approach on how to handle the special and rare case of 4 wiimote + 4 gc controllers. because the current approach contains a lot of clashes between P0-P3 and P4-P7 to be worth it. you should stay away from changing the poll callbacks and implement the logic somewhere else.

It might be a lot simpler to just limit them to port 4-7 in wii mode. or is there any specific reason why you would want to connect a gc controller with 4 or less players when you can connect a classic controller ?

RobertTheSable commented:
> it does look a lot better now :)
> 
> I think you need a new approach on how to handle the special and rare case of 4 wiimote + 4 gc controllers. because the current approach contains a lot of clashes between P0-P3 and P4-P7 to be worth it. you should stay away from changing the poll callbacks and implement the logic somewhere else.
> 
> It might be a lot simpler to just limit them to port 4-7 in wii mode. or is there any specific reason why you would want to connect a gc controller with 4 or less players when you can connect a classic controller ?

Smash Bros. Brawl comes to mind as an example of GC controllers being preferred for 1-4 players. I think it even uses the analog triggers, which might not be emulate-able with a CC instead? (I know the original has soft triggers, but the Pro doesn't, so I'm not sure if any game ever used them). There are also a few games that support GC controllers but not CCs, though the only one that comes to mind off the top of my head is the Metal Slug Collection. Oh, and there's some DDR games for the Wii that use a dance mat plugged into the GC ports.

I'm fine with switching GC to ports 4-7 only and defaulting to Wiimote(or sideways Wiimote, or Wiimote+Nunchuck as @bslenul suggested), whichever is preferred. I don't like the current behavior of leaving RetroPad unmapped though, since it leaves a controller option in the menu that's not usable, which is confusing from a UX perspective IMO.

> > Does it behave different in Port 1 vs. Port 5?
> 
> Ah yep, it works if I set my controller to port 5. Is that on purpose then? Seems a bit inconvenient if you have "Max Users" set to "4" or less in RetroArch :p

Yeah, using port 5+ for Gamecube was my goal for games like Bomberman Blast(which I  think might just be Bomberman Blast) that take advantage of Wii and Gamecube controller simultaneously.

aliaspider commented:
ok how about this : keep two lookup tables for wii and gc ports. if the input port is 0-3 try to give it that slot, for the other cases ( slot reserved or ports 4-7) just give it the first free slot in the table. if the table is full just clear the descriptors for that port and bail.

if it turns out to be too complex, just limit the gc controllers to ports 4-7 and be done with it. I think such a niche use-case isn't worth the hassle and code complexity that comes with it.

bslenul commented:
> Yeah, using port 5+ for Gamecube was my goal for games like Bomberman Blast(which I think might just be Bomberman Blast) that take advantage of Wii and Gamecube controller simultaneously.

Yeah sorry, I meant for using a GC controller as player 1 for Wii games, it just feels weird/confusing to have to set your controller to port 5 when port 1 already appears as a "GameCube Controller" in Controls menu :p

Commit:
Changes based on PR feedback:

- Disable port 1-4 and 5-8 aliasing for GC Controllers.
- Default to Wii mode in ports 1-4.

aliaspider commented on Source/Core/DolphinLibretro/Input.cpp:
```diff
@@ -655,48 +717,81 @@ void retro_set_controller_port_device(unsigned port, unsigned device)
+  }
+  all_descs.push_back({0});
+  Libretro::environ_cb(RETRO_ENVIRONMENT_SET_INPUT_DESCRIPTORS, &all_descs[0]);
 }
```
I think you forgot `RETRO_DEVICE_REAL_WIIMOTE` here.
you probably don't event need to duplicate the logic with an new switch block, it would actually make more sense to just keep the `all_descs` array static and update only one port at a time.
Edit: nvm, different desc array sizes I guess...

aliaspider commented on Source/Core/DolphinLibretro/Input.cpp:
```diff
@@ -464,22 +484,28 @@ void retro_set_input_state(retro_input_state_t cb)
-  retro_input_descriptor* desc = Libretro::Input::descGC;
+  Libretro::Input::RemoveDevicesForPort(port);
+  if (device != RETRO_DEVICE_NONE)
+    Libretro::Input::AddDevicesForPort(port);
```
you might want to check for `RETRO_DEVICE_REAL_WIIMOTE` here too.

aliaspider commented on Source/Core/DolphinLibretro/Input.cpp:
```diff
@@ -655,48 +717,81 @@ void retro_set_controller_port_device(unsigned port, unsigned device)
+      }
+      break;
+    }
+    for (int j = 0; desc[j].device != 0; j++)
```
just nitpicking, the test should be for `description` being NULL instead.

aliaspider commented:
can you do a rebase when you are done? that should get rid of the merge commit.

aliaspider commented on Source/Core/DolphinLibretro/Input.cpp:
```diff
@@ -495,36 +521,38 @@ void retro_set_controller_port_device(unsigned port, unsigned device)
+    gcTriggers->SetControlExpression(2,
+                                     "`" + devAnalog + ":Trigger0+`|L3");  // L-trigger Soft Press
+    gcTriggers->SetControlExpression(3,
+                                     "`" + devAnalog + ":Trigger1+`|R3");  // R-trigger Soft Press
```
mixed triggers should have a digital control expression for 0-1 and an analog one for 2-3. I don't see the point of using such a complex expression. also why `L3`/`R3` ? triggers are either `L`/`R` or `L2`/`R2`

aliaspider replied:
ok I assume this part here was the problem : https://github.com/libretro/dolphin/blob/master/Source/Core/InputCommon/ControllerEmu/ControlGroup/MixedTriggers.cpp#L54-L60
were the analog inputs being ignored when using digital input for 0-1 ?

aliaspider replied:
would changing the code in `MixedTriggers.cpp` like this work with the simpler control expressions? :
```// Apply threshold:
    if (button_value > threshold)
    {
      // Fully activate analog:
       if (analog_value == 0.0)
          analog_value = 1.0;

      // Activate button:
      *digital |= bitmasks[i];
    }

```
I don't have a suitable gamepad to test.

aliaspider replied:
regarding the L2/L3 part, I see you were tricking the game into thinking that it was a half press, by only setting the analog input to full while keeping the digital input cleared? kinda weird but I guess that is ok if it does work for the games you tested. L2/L3 aren't used for anything else afterall.

aliaspider commented on Source/Core/DolphinLibretro/Input.cpp:
```diff
@@ -541,40 +569,40 @@ void retro_set_controller_port_device(unsigned port, unsigned device)
+      ccTriggers->SetControlExpression(0, "`" + devAnalog + ":Trigger0+`");  // L-trigger
+      ccTriggers->SetControlExpression(1, "`" + devAnalog + ":Trigger1+`");  // R-trigger
+      ccTriggers->SetControlExpression(2, "`" + devAnalog + ":Trigger0+`");  // L-trigger Analog
+      ccTriggers->SetControlExpression(3, "`" + devAnalog + ":Trigger1+`");  // R-trigger Analog
```
same as above.

RobertTheSable replied:
Do you want me to use the same logic for GC triggers here? I didn't actually change the logic here.

aliaspider replied:
it used to be just `L` and `R` for '0' and '1'. but since the digital input would override the analog one, it does make sense to use only analog here.
you can keep it this way then, unless the changes I suggested above to `MixedTriggers.cpp` would work better.

Commit:
PR Review Fixes.

aliaspider commented on Source/Core/DolphinLibretro/Input.cpp:
```diff
@@ -602,24 +630,58 @@ void retro_set_controller_port_device(unsigned port, unsigned device)
+        wmButtons->SetControlExpression(2, "Start");                         // 1
+        wmButtons->SetControlExpression(3, "Select");                        // 2
+        wmButtons->SetControlExpression(4, "L");                             // -
+        wmButtons->SetControlExpression(5, "R");                             // +
```
maybe move the 0/1 expressions in here too? since you are overriding them anyway in the `else` block.

aliaspider commented on Source/Core/DolphinLibretro/Input.cpp:
```diff
@@ -464,22 +484,28 @@ void retro_set_input_state(retro_input_state_t cb)
   std::string devMouse = Libretro::Input::GetQualifiedName(port, RETRO_DEVICE_MOUSE);
   std::string devPointer = Libretro::Input::GetQualifiedName(port, RETRO_DEVICE_POINTER);
 #if 0
+  std::string devMouse = Libretro::Input::GetQualifiedName(port, RETRO_DEVICE_MOUSE);
```
you missed this one.

RitschiRitsch commented:
May it be possible to approve and merge this PR? The missing analog triggers are game breaking for Super Mario Sunshine. Thank you.

teconmoon commented:
Any progress on this issue? Seems like activity has stalled out for some time.

Same for me, Super Mario Sunshine is the game that breaks the worst with this, but some other games like Mario Party 4 suffer from minor issues like being unable to open the map without an analog trigger (does have a gecko code workaround).

RobertTheSable commented:
@aliaspider what's blocking this PR from being ready to merge?

Papermanzero commented:
Can someone make a test build? This is one of the most useful PR‘s.

bslenul commented:
https://github.com/bslenul/dolphin/releases/tag/input-fix :)

Papermanzero commented:
The tilt function is not working for the wii mote. for whatever reason the tilt is always doing a right shift independent of the actual input

inactive123 commented:
It's merged. We have to keep moving this core forward.
